server/routerlicious/packages/services/src/mongodb.ts

```diff
@@ -7,103 +7,174 @@ import { assert } from "console";
 import * as core from "@fluidframework/server-services-core";
 import { AggregationCursor, Collection, MongoClient, MongoClientOptions } from "mongodb";
 import { Lumberjack } from "@fluidframework/server-services-telemetry";
+import { requestWithRetry } from "@fluidframework/server-services-core";
+import { MongoErrorRetryAnalyzer } from "./mongoExceptionRetryRules";
 
 const MaxFetchSize = 2000;
 
-export class MongoCollection<T> implements core.ICollection<T> {
-    constructor(private readonly collection: Collection<T>) {
+export class MongoCollection<T> implements core.ICollection<T>, core.IRetryAble {
+    public readonly retryEnabled: boolean;
+    constructor(
+        private readonly collection: Collection<T>,
+        private readonly clientSideRetry?: boolean,
+    ) {
+        assert(collection, "collection must be defined");
+        this.retryEnabled = clientSideRetry ?? false;
     }
 
-    public aggregate(pipeline: any, options?: any): AggregationCursor<T> {
-        return this.collection.aggregate(pipeline, options);
+    public async aggregate(pipeline: any, options?: any): Promise<AggregationCursor<T>> {
+        const req = async () => new Promise<AggregationCursor<T>>(
+            (resolve) => resolve(this.collection.aggregate(pipeline, options)));
+        return this.requestWithRetry(req, "MongoCollection.aggregate");
     }
 
-    // eslint-disable-next-line @typescript-eslint/ban-types,@typescript-eslint/promise-function-async
-    public find(query: object, sort: any, limit = MaxFetchSize, skip?: number): Promise<T[]> {
-        let queryCursor = this.collection
-            .find(query)
-            .sort(sort)
-            .limit(limit);
-
-        if (skip) {
-            queryCursor = queryCursor.skip(skip);
-        }
-        return queryCursor.toArray();
+    public async find(query: object, sort: any, limit = MaxFetchSize, skip?: number): Promise<T[]> {
+        const req: () => Promise<T[]> = async () => {
+            let queryCursor = this.collection
+                .find(query)
+                .sort(sort)
+                .limit(limit);
+
+            if (skip) {
+                queryCursor = queryCursor.skip(skip);
+            }
+            return queryCursor.toArray();
+        };
+        return this.requestWithRetry(req, "MongoCollection.find");
     }
 
-    // eslint-disable-next-line @typescript-eslint/ban-types,@typescript-eslint/promise-function-async
```

Comment:
Interesting that we had all these "disable lint rules" defined here and in this file. I wonder why the original implementation did not use the `async` keyword, even though most of these functions returned promises. Does adding the `async` keyword, like you did, cause any side effect?

Reply:
I don't know to why it was added in the first place and what could be side effect now. I still think if we use promise, we should async all the way. An idea in c#